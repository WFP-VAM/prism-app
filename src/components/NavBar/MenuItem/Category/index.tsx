import React from 'react';
import {
  Switch,
  Typography,
  withStyles,
  createStyles,
  WithStyles,
} from '@material-ui/core';

import { LayerType } from '../../../../config/types';

function Category({ classes, title, layers }: CategoryProps) {
  return (
    <div className={classes.container}>
      <Typography variant="body2" className={classes.title}>
        {title}
      </Typography>
      <hr />
      {layers.map(({ id, title: layersTitle }) => (
        <div key={id} className={classes.listContainer}>
          <Switch
            size="small"
            color="default"
            inputProps={{ 'aria-label': 'checkbox with default color' }}
          />{' '}
          <Typography variant="body1">{layersTitle}</Typography>
        </div>
      ))}
    </div>
  );
}

const styles = () =>
  createStyles({
    container: {
      marginBottom: 16,
    },

    listContainer: {
      display: 'flex',
      marginBottom: 8,
    },

    title: {
      fontWeight: 'bold',
      textAlign: 'left',
    },
  });

export interface CategoryProps extends LayerType, WithStyles<typeof styles> {}

export default withStyles(styles)(Category);
