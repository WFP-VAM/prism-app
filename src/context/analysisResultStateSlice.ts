import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';
import { Position, FeatureCollection, Feature } from 'geojson';
import moment from 'moment';
import { get } from 'lodash';
import type { CreateAsyncThunkTypes, RootState } from './store';
import { defaultBoundariesFile } from '../config';
import {
  AggregationOperations,
  AsyncReturnType,
  BoundaryLayerProps,
  NSOLayerProps,
  ThresholdDefinition,
  WMSLayerProps,
  WfsRequestParams,
  LayerKey,
  ExposedPopulationDefinition,
  TableType,
} from '../config/types';
import {
  BaselineLayerResult,
  AnalysisResult,
  ApiData,
  BaselineLayerData,
  checkBaselineDataLayer,
  fetchApiData,
  generateFeaturesFromApiData,
  createLegendFromFeatureArray,
  KeyValueResponse,
  scaleAndFilterAggregateData,
  ExposedPopulationResult,
  scaleFeatureStat,
} from '../utils/analysis-utils';
import { getWCSLayerUrl } from './layers/wms';
import { getBoundaryLayerSingleton, LayerDefinitions } from '../config/utils';
import { Extent } from '../components/MapView/Layers/raster-utils';
import { layerDataSelector } from './mapStateSlice/selectors';
import { LayerData, LayerDataParams, loadLayerData } from './layers/layer-data';
import { DataRecord, NSOLayerData } from './layers/nso';
import { BoundaryLayerData } from './layers/boundary';
import { isLocalhost } from '../serviceWorker';

const ANALYSIS_API_URL = 'https://prism-api.ovio.org/stats'; // TODO both needs to be stored somewhere

export type TableRowType = { [key: string]: string | number };
export type TableData = {
  columns: string[];
  rows: TableRowType[];
};

type AnalysisResultState = {
  definition?: TableType;
  tableData?: TableData;
  result?: AnalysisResult;
  error?: string;
  isLoading: boolean;
  isMapLayerActive: boolean;
  isDataTableDrawerActive: boolean;
  isExposureLoading: boolean;
};

export type TableRow = {
  localName: string;
  name: string;
  baselineValue: DataRecord['value'];
  coordinates?: Position;
} & { [k in AggregationOperations]?: number };

const initialState: AnalysisResultState = {
  isLoading: false,
  isMapLayerActive: true,
  isDataTableDrawerActive: false,
  isExposureLoading: false,
};

/* Gets a public URL for the admin boundaries used by this application.
 *
 * If the application is in development, localhost is not accessible publicly.
 * Therefore, we will return a pre-set constant to be used for development.
 *
 * If the application is in production, we will attempt to construct a public URL that the backend should be able to access.
 */
function getAdminBoundariesURL() {
  const adminBoundariesPath = getBoundaryLayerSingleton().path;
  // already a remote location, so return it.
  if (adminBoundariesPath.startsWith('http')) {
    return adminBoundariesPath;
  }
  // do not send a local path to the API, use a fixed boundary file instead.
  if (isLocalhost) {
    return defaultBoundariesFile;
  }
  // the regex here removes the dot at the beginning of a path, if there is one.
  // e.g the path might be ' ./data/xxx '  instead of ' /data/xxx '
  return window.location.origin + adminBoundariesPath.replace(/^\./, '');
}

function generateTableFromApiData(
  statistic: AggregationOperations,
  aggregateData: AsyncReturnType<typeof fetchApiData>, // data from api
  // admin layer, both props and data. Aimed to closely match LayerData<BoundaryLayerProps>
  {
    layer: adminLayer,
    data: adminLayerData,
  }: { layer: BoundaryLayerProps; data: BoundaryLayerData },
  // baseline layer, both props and data
  {
    layer: { adminLevel },
    data: { layerData: baselineLayerData },
  }: { layer: NSOLayerProps; data: NSOLayerData },
): TableRow[] {
  // find the key that will let us reference the names of the bounding boxes. We get the one corresponding to the specific level of baseline, or the first if we fail.
  const adminLevelName =
    adminLayer.adminLevelNames[adminLevel - 1] || adminLayer.adminLevelNames[0];
  // for local name too.
  const adminLevelLocalName =
    adminLayer.adminLevelLocalNames[adminLevel - 1] ||
    adminLayer.adminLevelLocalNames[0];

  return (aggregateData as KeyValueResponse[]).map(row => {
    // find feature (a cell on the map) from admin boundaries json that closely matches this api row.
    // we decide it matches if the feature json has the same name as the name for this row.
    // once we find it we can get the corresponding local name.

    const featureBoundary = adminLayerData.features.find(
      feature => feature.properties?.[adminLevelName] === row[adminLevelName],
    );

    const name: string =
      featureBoundary?.properties?.[adminLevelName] || 'No Name';
    const localName: string =
      featureBoundary?.properties?.[adminLevelLocalName] || 'No Name';

    // we are searching the data of baseline layer to find the data associated with this feature
    // adminKey here refers to a specific feature (could be several) where the data is attached to.
    const baselineValue =
      baselineLayerData.find(({ adminKey }) => {
        // we only check startsWith because the adminCode grows longer the deeper the level.
        // For example, 34 is state and 14 is district, therefore 3414 is a specific district in a specific state.
        // if this baseline layer only focuses on a higher level (just states) it would only contain 34, but every feature is very specific (uses the full number 3414)
        // therefore checking the start will cover all levels.
        return featureBoundary?.properties?.[adminLayer.adminCode].startsWith(
          adminKey,
        );
      })?.value || 'No Data';

    const tableRow: TableRow = {
      name,
      localName,
      [statistic]: get(row, statistic, 0),
      baselineValue,
      coordinates: (featureBoundary?.geometry as any).coordinates[0][0][0], // TODO likely will not keep
    };
    return tableRow;
  });
}

export type AnalysisDispatchParams = {
  baselineLayer: NSOLayerProps;
  hazardLayer: WMSLayerProps;
  extent: Extent;
  threshold: ThresholdDefinition;
  date: ReturnType<Date['getTime']>; // just a hint to developers that we give a date number here, not just any number
  statistic: AggregationOperations; // we might have to deviate from this if analysis accepts more than what this enum provides
  isExposure: boolean;
};

export type ExposedPopulationDispatchParams = {
  exposure: ExposedPopulationDefinition;
  statistic: AggregationOperations;
  extent: Extent;
  date: ReturnType<Date['getTime']>;
  wfsLayerId: LayerKey;
};

const createAPIRequestParams = (
  geotiffLayer: WMSLayerProps,
  extent: Extent,
  date: ReturnType<Date['getTime']>,
  params: WfsRequestParams | NSOLayerProps,
): ApiData => {
  const adminBoundaries = getBoundaryLayerSingleton();

  const groupBy =
    adminBoundaries.adminLevelNames[(params as NSOLayerProps).adminLevel - 1] ||
    adminBoundaries.adminLevelNames[0];

  const wfsParams = (params as WfsRequestParams).layer_name
    ? { wfs_params: params as WfsRequestParams }
    : undefined;

  // we force group_by to be defined with &
  // eslint-disable-next-line camelcase
  const apiRequest: ApiData = {
    geotiff_url: getWCSLayerUrl({
      layer: geotiffLayer,
      date: geotiffLayer.wcsConfig?.timeSupport === true ? date : undefined,
      extent,
    }),
    zones_url: getAdminBoundariesURL(),
    group_by: groupBy,
    ...wfsParams,
  };

  return apiRequest;
};

export const requestAndStoreExposedPopulation = createAsyncThunk<
  AnalysisResult,
  ExposedPopulationDispatchParams,
  CreateAsyncThunkTypes
>('analysisResultState/requestAndStoreExposedPopulation', async params => {
  const { exposure, date, extent, statistic, wfsLayerId } = params;

  const { id, key } = exposure;

  const wfsLayer = LayerDefinitions[wfsLayerId] as WMSLayerProps;
  const populationLayer = LayerDefinitions[id] as WMSLayerProps;

  const wfsParams: WfsRequestParams = {
    url: `${wfsLayer.baseUrl}/ows`,
    layer_name: wfsLayer.serverLayerName,
    time: moment(date).format('YYYY-MM-DD'),
    key,
  };

  const apiRequest = createAPIRequestParams(
    populationLayer,
    extent,
    date,
    wfsParams,
  );

  const apiFeatures = (await fetchApiData(
    ANALYSIS_API_URL,
    apiRequest,
  )) as Feature[];

  const { scale, offset } = populationLayer.wcsConfig ?? {
    scale: undefined,
    offset: undefined,
  };

  const features =
    !scale && !offset
      ? apiFeatures
      : apiFeatures.map(f =>
          scaleFeatureStat(f, scale || 1, offset || 0, statistic),
        );

  const collection: FeatureCollection = {
    type: 'FeatureCollection',
    features,
  };

  const legend = createLegendFromFeatureArray(features, statistic);
  const legendText = wfsLayer.title;

  return new ExposedPopulationResult(
    collection,
    statistic,
    legend,
    legendText,
    key,
  );
});

export const requestAndStoreAnalysis = createAsyncThunk<
  AnalysisResult,
  AnalysisDispatchParams,
  CreateAsyncThunkTypes
>('analysisResultState/requestAndStoreAnalysis', async (params, api) => {
  const {
    hazardLayer,
    date,
    baselineLayer,
    extent,
    statistic,
    threshold,
  } = params;
  const baselineData = layerDataSelector(baselineLayer.id)(
    api.getState(),
  ) as LayerData<NSOLayerProps>;
  const adminBoundaries = getBoundaryLayerSingleton();
  const adminBoundariesData = layerDataSelector(adminBoundaries.id)(
    api.getState(),
  ) as LayerData<BoundaryLayerProps>;

  if (!adminBoundariesData) {
    throw new Error('Boundary Layer not loaded!');
  }
  const apiRequest = createAPIRequestParams(
    hazardLayer,
    extent,
    date,
    baselineLayer,
  );

  const aggregateData = scaleAndFilterAggregateData(
    await fetchApiData(ANALYSIS_API_URL, apiRequest),
    hazardLayer,
    statistic,
    threshold,
  );
  let loadedAndCheckedBaselineData: BaselineLayerData;
  // if the baselineData doesn't exist, lets load it, otherwise check then load existing data.
  // similar code can be found at impact.ts
  if (!baselineData) {
    const {
      payload: { data },
    } = (await api.dispatch(
      loadLayerData({ layer: baselineLayer, extent } as LayerDataParams<
        NSOLayerProps
      >),
    )) as { payload: { data: unknown } };

    // eslint-disable-next-line fp/no-mutation
    loadedAndCheckedBaselineData = checkBaselineDataLayer(
      baselineLayer.id,
      data,
    );
  } else {
    // eslint-disable-next-line fp/no-mutation
    loadedAndCheckedBaselineData = checkBaselineDataLayer(
      baselineLayer.id,
      baselineData.data,
    );
  }

  const features = generateFeaturesFromApiData(
    aggregateData,
    loadedAndCheckedBaselineData,
    apiRequest.group_by,
    statistic,
  );

  const tableRows: TableRow[] = generateTableFromApiData(
    statistic,
    aggregateData,
    adminBoundariesData,
    { layer: baselineLayer, data: loadedAndCheckedBaselineData },
  );

  return new BaselineLayerResult(
    tableRows,
    {
      ...adminBoundariesData.data,
      features,
    },
    hazardLayer,
    baselineLayer,
    statistic,
    threshold,
    // we never use the raw api data besides for debugging. So lets not bother saving it in Redux for production
    process.env.NODE_ENV === 'production' ? undefined : aggregateData,
  );
});

export const analysisResultSlice = createSlice({
  name: 'analysisResultState',
  initialState,
  reducers: {
    addTableData: (state, { payload }: PayloadAction<TableData>) => ({
      ...state,
      tableData: payload,
    }),
    setIsMapLayerActive: (state, { payload }: PayloadAction<boolean>) => ({
      ...state,
      isMapLayerActive: payload,
    }),
    setIsDataTableDrawerActive: (
      state,
      { payload }: PayloadAction<boolean>,
    ) => ({
      ...state,
      isDataTableDrawerActive: payload,
    }),
    setCurrentDataDefinition: (
      state,
      { payload }: PayloadAction<TableType>,
    ) => ({
      ...state,
      definition: payload,
    }),
    hideDataTableDrawer: state => ({
      ...state,
      isDataTableDrawerActive: false,
    }),
    clearAnalysisResult: state => ({
      ...state,
      result: undefined,
    }),
  },
  extraReducers: builder => {
    builder.addCase(
      requestAndStoreExposedPopulation.fulfilled,
      (
        { result, ...rest },
        { payload }: PayloadAction<AnalysisResult>,
      ): AnalysisResultState => ({
        ...rest,
        result: payload as ExposedPopulationResult,
        isExposureLoading: false,
      }),
    );

    builder.addCase(
      requestAndStoreExposedPopulation.rejected,
      (state, action): AnalysisResultState => ({
        ...state,
        isExposureLoading: false,
        error: action.error.message
          ? action.error.message
          : action.error.toString(),
      }),
    );

    builder.addCase(
      requestAndStoreExposedPopulation.pending,
      (state): AnalysisResultState => ({
        ...state,
        isExposureLoading: true,
      }),
    );

    builder.addCase(
      requestAndStoreAnalysis.fulfilled,
      (
        { result, ...rest },
        { payload }: PayloadAction<AnalysisResult>,
      ): AnalysisResultState => ({
        ...rest,
        isLoading: false,
        result: payload as BaselineLayerResult,
      }),
    );

    builder.addCase(
      requestAndStoreAnalysis.rejected,
      (state, action): AnalysisResultState => ({
        ...state,
        isLoading: false,
        error: action.error.message
          ? action.error.message
          : action.error.toString(),
      }),
    );

    builder.addCase(
      requestAndStoreAnalysis.pending,
      (state): AnalysisResultState => ({
        ...state,
        isLoading: true,
      }),
    );
  },
});

// Getters
export const getCurrentDefinition = (state: RootState): TableType | undefined =>
  state.analysisResultState.definition;

export const getCurrentData = (state: RootState): TableData =>
  state.analysisResultState.tableData || { columns: [], rows: [] };

export const analysisResultSelector = (
  state: RootState,
): AnalysisResult | undefined => state.analysisResultState.result;

export const isAnalysisLoadingSelector = (state: RootState): boolean =>
  state.analysisResultState.isLoading;

export const isExposureAnalysisLoadingSelector = (state: RootState): boolean =>
  state.analysisResultState.isExposureLoading;

export const isAnalysisLayerActiveSelector = (state: RootState): boolean =>
  state.analysisResultState.isMapLayerActive;

export const isDataTableDrawerActiveSelector = (state: RootState): boolean =>
  state.analysisResultState.isDataTableDrawerActive;

// Setters
export const {
  addTableData,
  setIsMapLayerActive,
  setIsDataTableDrawerActive,
  setCurrentDataDefinition,
  hideDataTableDrawer,
  clearAnalysisResult,
} = analysisResultSlice.actions;

export default analysisResultSlice.reducer;
