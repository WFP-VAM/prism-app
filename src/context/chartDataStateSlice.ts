import moment from 'moment';
import { createSlice, PayloadAction, createAsyncThunk } from '@reduxjs/toolkit';
import * as Papa from 'papaparse';
import type { CreateAsyncThunkTypes, RootState } from './store';
import { TableData } from './tableStateSlice';

type DatasetState = {
  data?: TableData;
  title?: string;
};

const initialState: DatasetState = {};

export type DatasetParams = {
  id: string;
  filepath: string;
};

export const loadDataset = createAsyncThunk<
  TableData,
  DatasetParams,
  CreateAsyncThunkTypes
>('datasetState/loadDataset', async (params: DatasetParams) => {
  const url = process.env.PUBLIC_URL + params.filepath;

  return new Promise<TableData>((resolve, reject) =>
    Papa.parse(url, {
      header: true,
      download: true,
      complete: results => {
        const row = results.data.find(item => item.Admin2_Code === params.id);

        return resolve({
          rows: [...results.data.slice(0, 1), row],
          columns: Object.keys(row),
        });
      },
      error: error => reject(error),
    }),
  );
});

export const datasetResultStateSlice = createSlice({
  name: 'DatasetResultSlice',
  initialState,
  reducers: {
    addEwsDataset: (
      { ...rest },
      { payload }: PayloadAction<TableData>,
    ): DatasetState => {
      const { rows, columns } = payload;
      const formattedRows = [
        Object.fromEntries(
          Object.keys(rows[0]).map(k => [
            k,
            k === 'level'
              ? rows[0][k]
              : moment(rows[0][k]).local().format('DD/mm/YYYY HH:MM'),
          ]),
        ),
        ...rows.slice(1, rows.length),
      ];

      return { ...rest, data: { rows: formattedRows, columns } };
    },
    addPointTitle: ({ ...rest }, { payload }: PayloadAction<string>) => {
      return { ...rest, title: payload };
    },
  },
  extraReducers: builder => {
    builder.addCase(
      loadDataset.fulfilled,
      ({ ...rest }, { payload }: PayloadAction<TableData>): DatasetState => ({
        ...rest,
        data: payload,
      }),
    );
  },
});

export const DatasetSelector = (state: RootState): TableData | undefined =>
  state.datasetState.data;
export const PointTitleSelector = (state: RootState): string | undefined =>
  state.datasetState.title;

// Setters
export const { addEwsDataset, addPointTitle } = datasetResultStateSlice.actions;

export default datasetResultStateSlice.reducer;
