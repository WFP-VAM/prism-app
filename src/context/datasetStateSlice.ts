import moment from 'moment';
import { orderBy } from 'lodash';
import { createSlice, PayloadAction, createAsyncThunk } from '@reduxjs/toolkit';
import type { CreateAsyncThunkTypes, RootState } from './store';
import { TableData } from './tableStateSlice';
import { ChartType } from '../config/types';

type BoundaryPropsDict = { [key: string]: BoundaryProps };

type ServerParams = {
  url: string;
  layerName: string;
};

type DatasetState = {
  data?: TableData;
  isLoading: boolean;
  adminBoundaryParams?: AdminBoundaryParams;
  id?: string;
};

const initialState: DatasetState = { isLoading: false };

type BoundaryProps = {
  code: number;
  urlPath: string;
  name: string;
};

export type AdminBoundaryParams = {
  boundaryProps: BoundaryPropsDict;
  serverParams: ServerParams;
  title: string;
  chartType: ChartType;
};

export type DatasetParams = {
  id: string;
  boundaryProps: BoundaryPropsDict;
  url: string;
  serverLayerName: string;
  selectedDate: number;
};

type DataItem = {
  date: number;
  value: number;
};

const getDatasetFromUrl = async (
  year: number,
  params: DatasetParams,
  startDate: number,
  endDate: number,
): Promise<DataItem[]> => {
  const { serverLayerName, url, id, boundaryProps } = params;

  const { code: adminCode, urlPath } = boundaryProps[id];

  const serverUrl = `${url}/${urlPath}/${year}.json`;

  const resp = await fetch(serverUrl);
  const results = await resp.json();

  const filteredRows = results.DataList.filter(
    (item: any) => item[id] === adminCode,
  )
    .map((item: any) => ({
      date: moment(item.time).valueOf(),
      value: item[serverLayerName],
    }))
    .filter(({ date }: DataItem) => date >= startDate && date <= endDate);

  return filteredRows;
};

export const loadDataset = createAsyncThunk<
  TableData,
  DatasetParams,
  CreateAsyncThunkTypes
>('datasetState/loadDataset', async (params: DatasetParams) => {
  const endDate = moment(params.selectedDate);
  const startDate = endDate.clone().subtract(1, 'year');

  const years = [endDate.year(), startDate.year()];

  const promises = years.map(year =>
    getDatasetFromUrl(year, params, startDate.valueOf(), endDate.valueOf()),
  );
  const resultsAll = await Promise.all(promises);

  const results: DataItem[] = resultsAll.reduce(
    (acc, item) => [...acc, ...item],
    [],
  );

  const sortedRows = orderBy(results, item => item.date).map((item, index) => ({
    ...item,
    day: `d${index + 1}`,
  }));

  const datesRows = sortedRows.reduce(
    (acc, obj) => ({
      ...acc,
      [obj.day]: moment(obj.date).format('YYYY-MM-DD'),
    }),
    {},
  );

  const valuesRows = sortedRows.reduce((acc, obj) => {
    if (!obj.value) {
      return acc;
    }

    return { ...acc, [obj.day]: obj.value.toString() };
  }, {});

  const columns = Object.keys(valuesRows);
  const data: TableData = {
    rows: [datesRows, valuesRows],
    columns,
  };

  return new Promise<TableData>(resolve => resolve(data));
});

export const datasetResultStateSlice = createSlice({
  name: 'DatasetResultSlice',
  initialState,
  reducers: {
    clearDataset: (): DatasetState => initialState,
    setBoundaryParams: (
      state,
      { payload }: PayloadAction<AdminBoundaryParams>,
    ): DatasetState => ({
      ...state,
      adminBoundaryParams: payload,
    }),
    updateAdminId: (
      state,
      { payload }: PayloadAction<string>,
    ): DatasetState => ({
      ...state,
      id: payload,
    }),
  },
  extraReducers: builder => {
    builder.addCase(
      loadDataset.fulfilled,
      ({ ...rest }, { payload }: PayloadAction<TableData>): DatasetState => ({
        ...rest,
        data: payload,
        isLoading: false,
      }),
    );

    builder.addCase(loadDataset.pending, state => ({
      ...state,
      isLoading: true,
    }));
  },
});

export const datasetSelector = (state: RootState): DatasetState =>
  state.datasetState;
export const loadingDatasetSelector = (state: RootState): boolean =>
  state.datasetState.isLoading;

// Setters
export const {
  clearDataset,
  setBoundaryParams,
  updateAdminId,
} = datasetResultStateSlice.actions;

export default datasetResultStateSlice.reducer;
