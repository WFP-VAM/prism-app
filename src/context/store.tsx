import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';
import filtersReduce from './filters/filtersSlice';

export const store = configureStore({
  reducer: {
    filters: filtersReduce,
  },
  // TODO: Instead of snoozing this check, we might want to
  // serialize the state
  middleware: getDefaultMiddleware({
    serializableCheck: false,
  }),
});

export type RootState = ReturnType<typeof store.getState>;
