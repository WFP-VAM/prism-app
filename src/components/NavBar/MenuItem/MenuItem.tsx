import React, { useState, useRef } from 'react';
import {
  Button,
  Typography,
  Popper,
  ClickAwayListener,
  Paper,
  withStyles,
  WithStyles,
  createStyles,
  Theme,
} from '@material-ui/core';

import Category from './Category/Category';
import { CategoryType } from '../../../config/types';

function MenuItem({ classes, title, icon, layersList }: MenuItemProps) {
  const [open, setOpen] = useState(false);
  const anchorRef = useRef<any>();

  return (
    <>
      <Button
        ref={anchorRef}
        className={classes.title}
        onClick={() => setOpen(prevOpen => !prevOpen)}
        aria-controls={open ? 'menu-list-grow' : undefined}
        aria-haspopup="true"
      >
        <img className={classes.icon} src={icon} alt={title} />
        <Typography variant="body2">{title}</Typography>
      </Button>

      <Popper
        open={open}
        anchorEl={anchorRef.current}
        className={classes.popper}
        placement="bottom-start"
        transition
        disablePortal
      >
        <ClickAwayListener onClickAway={() => setOpen(false)}>
          <Paper className={classes.paper}>
            {layersList.map(layers => (
              <Category key={layers.title} {...layers} />
            ))}
          </Paper>
        </ClickAwayListener>
      </Popper>
    </>
  );
}

const styles = (theme: Theme) =>
  createStyles({
    title: {
      margin: '0px 14px',
      textTransform: 'uppercase',
      color: 'inherit',

      '&::after': {
        content: '""',
        display: 'inline-block',
        width: 0,
        height: 0,
        marginLeft: 3.5,
        verticalAlign: 3.5,
        borderTop: '3.5px solid',
        borderRight: '3.5px solid transparent',
        borderBottom: 0,
        borderLeft: '3.5px solid transparent',
      },
    },

    icon: {
      width: 18,
      marginRight: 6,
    },

    popper: {
      marginTop: 8,
    },

    paper: {
      padding: '8px 16px',
      backgroundColor: `${theme.palette.primary.main}f9`,
      borderRadius: 4,
    },
  });

export interface MenuItemProps
  extends CategoryType,
    WithStyles<typeof styles> {}

export default withStyles(styles)(MenuItem);
