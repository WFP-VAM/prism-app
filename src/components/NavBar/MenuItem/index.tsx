import React, { useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import {
  Button,
  Typography,
  Popover,
  withStyles,
  WithStyles,
  createStyles,
  Theme,
  Switch,
} from '@material-ui/core';

import { MenuItemType, LayerType, TableType } from '../../../config/types';
import { addLayer, removeLayer } from '../../../context/mapStateSlice';
import { loadTable } from '../../../context/tableStateSlice';
import { layersSelector } from '../../../context/mapStateSlice/selectors';

function MenuItem({ classes, title, icon, layersCategories }: MenuItemProps) {
  const [anchorEl, setAnchorEl] = useState<HTMLButtonElement | null>(null);

  const selectedLayers = useSelector(layersSelector);
  const dispatch = useDispatch();

  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    setAnchorEl(event.currentTarget);
  };

  const handleClose = () => {
    setAnchorEl(null);
  };

  const toggleLayerValue = (prevChecked: boolean, layer: LayerType) => {
    if (prevChecked) {
      dispatch(removeLayer(layer));
    } else {
      dispatch(addLayer(layer));
    }
  };

  const showTableClicked = (table: TableType) => {
    dispatch(loadTable(table.id));
  };

  const open = Boolean(anchorEl);
  const id = open ? 'menu-item-popover' : undefined;

  return (
    <>
      <Button
        className={classes.title}
        onClick={handleClick}
        aria-describedby={id}
      >
        <img className={classes.icon} src={icon} alt={title} />
        <Typography variant="body2">{title}</Typography>
      </Button>

      <Popover
        id={id}
        open={open}
        anchorEl={anchorEl}
        onClose={handleClose}
        className={classes.popover}
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'center',
        }}
        transformOrigin={{
          vertical: 'top',
          horizontal: 'center',
        }}
        PaperProps={{
          className: classes.paper,
        }}
      >
        {layersCategories.map(({ title: categoryTitle, layers, tables }) => (
          <div key={categoryTitle} className={classes.categoryContainer}>
            <Typography variant="body2" className={classes.categoryTitle}>
              {categoryTitle}
            </Typography>
            <hr />

            {layers.map(layer => {
              const { id: layerId, title: layerTitle } = layer;
              const selected = Boolean(
                selectedLayers.find(({ id: testId }) => testId === layerId),
              );
              return (
                <div key={layerId} className={classes.layersContainer}>
                  <Switch
                    size="small"
                    color="default"
                    checked={selected}
                    onChange={() => toggleLayerValue(selected, layer)}
                    inputProps={{ 'aria-label': layerTitle }}
                  />{' '}
                  <Typography variant="body1">{layerTitle}</Typography>
                </div>
              );
            })}

            {tables.map(table => (
              <Button
                className={classes.button}
                id={table.title}
                key={table.title}
                onClick={() => showTableClicked(table)}
              >
                <Typography variant="body1">{table.title}</Typography>
              </Button>
            ))}
          </div>
        ))}
      </Popover>
    </>
  );
}

const styles = (theme: Theme) =>
  createStyles({
    button: {
      textTransform: 'none',
    },

    title: {
      margin: '0px 14px',
      textTransform: 'uppercase',
      color: 'inherit',

      '&::after': {
        content: '""',
        display: 'inline-block',
        width: 0,
        height: 0,
        marginLeft: 3.5,
        verticalAlign: 3.5,
        borderTop: '3.5px solid',
        borderRight: '3.5px solid transparent',
        borderBottom: 0,
        borderLeft: '3.5px solid transparent',
      },
    },

    icon: {
      width: 18,
      marginRight: 6,
    },

    popover: {
      marginTop: 8,
    },

    paper: {
      padding: '8px 16px',
      backgroundColor: `${theme.palette.primary.main}f9`,
      borderRadius: theme.shape.borderRadius,
    },

    categoryContainer: {
      marginBottom: 16,
    },

    categoryTitle: {
      fontWeight: 'bold',
      textAlign: 'left',
    },

    layersContainer: {
      display: 'flex',
      marginBottom: 8,
    },
  });

export interface MenuItemProps
  extends MenuItemType,
    WithStyles<typeof styles> {}

export default withStyles(styles)(MenuItem);
