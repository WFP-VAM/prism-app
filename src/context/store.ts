import {
  combineReducers,
  configureStore,
  getDefaultMiddleware,
} from '@reduxjs/toolkit';

import mapStateReduce from './mapStateSlice';
import serverStateReduce from './serverStateSlice';
import tableStateReduce from './tableStateSlice';
import tooltipStateReduce from './tooltipStateSlice';
import notificationStateReduce, {
  errorToNotificationMiddleware,
} from './notificationStateSlice';
import analysisResultStateSlice from './analysisResultStateSlice';

const reducer = combineReducers({
  mapState: mapStateReduce,
  serverState: serverStateReduce,
  tableState: tableStateReduce,
  tooltipState: tooltipStateReduce,
  analysisResultState: analysisResultStateSlice,
  notificationState: notificationStateReduce,
});

export const store = configureStore({
  reducer,
  // TODO: Instead of snoozing this check, we might want to
  // serialize the state
  middleware: getDefaultMiddleware({
    serializableCheck: false,
    immutableCheck: {
      ignoredPaths: ['mapState.layersData', 'analysisResultState.result'],
    },
  }).concat(errorToNotificationMiddleware),
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof reducer>;

export type ThunkApi = {
  dispatch: AppDispatch;
  getState: () => RootState;
  // can add 'extra' and 'rejectValue' here if we ever need them
  // https://redux-toolkit.js.org/usage/usage-with-typescript#createasyncthunk
};

// Used as the third type definition for typing createAsyncThunk
// e.g. createAsyncThunk<ReturnType, HandlerType, CreateAsyncThunkTypes>
export type CreateAsyncThunkTypes = {
  dispatch: AppDispatch;
  state: RootState;
};
