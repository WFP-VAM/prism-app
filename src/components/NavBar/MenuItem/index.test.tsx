import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';

import MenuItem from '.';
import { store } from '../../../context/store';

const props = {
  title: 'title',
  icon: 'icon.png',
  layersCategories: [
    {
      title: 'Category 1',
      layers: [
        {
          id: 'ID',
          title: 'layer',
          type: 'wms',
          hasDate: false,
          opacity: 0.5,
          legendText: 'legendText',
        },
      ],
      tables: [],
    },
  ],
};

test('renders as expected', () => {
  const { container } = render(
    <Provider store={store}>
      <MenuItem {...props} />
    </Provider>,
  );
  expect(container).toMatchSnapshot();
});
