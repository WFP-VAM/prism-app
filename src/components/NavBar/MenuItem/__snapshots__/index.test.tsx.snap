// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<div>
  <mock-button
    classname="MenuItem-title-2"
  >
    <img
      alt="title"
      class="MenuItem-icon-3"
      src="icon.png"
    />
    <mock-typography
      variant="body2"
    >
      title
    </mock-typography>
  </mock-button>
  <mock-popover
    anchororigin="[object Object]"
    classname="MenuItem-popover-4"
    open="false"
    paperprops="[object Object]"
    transformorigin="[object Object]"
  >
    <div
      class="MenuItem-categoryContainer-6"
    >
      <mock-typography
        classname="MenuItem-categoryTitle-7"
        variant="body2"
      >
        Category 1
      </mock-typography>
      <hr />
      <div
        class="MenuItem-layersContainer-8"
      >
        <mock-switch
          checked="false"
          color="default"
          inputprops="[object Object]"
          size="small"
        />
         
        <mock-typography
          variant="body1"
        >
          layer
        </mock-typography>
      </div>
    </div>
  </mock-popover>
</div>
`;
